rfcs/syntax-floor-division-operator.md

```diff
@@ -0,0 +1,62 @@
+
+A typical mistake is to forget to use `math.floor`. This can produce subtle issues ranging from slightly wrong results to script errors. A script error could occur, for example, when the result of division is used to fetch from a table with only integer keys, which produces nil and a script error happens soon after. Another type of error occurs when an accidental fractional number is passed to a C function. Depending on the implementation, the C function could raise an error (if it checks that the number is actually an integer) or cause logic errors due to rounding.
+
+Particularly problematic is incorrect code which seems to work with frequently used data, only to fail with some rare input. For example, image sizes often have power of two dimensions, so code dealing with them may appear to work fine until much later some rare image has an odd size and a division by two in the code does not produce the correct result. Due to better ergonomics of the floor division operator, it becomes a second nature to write `//` everywhere when integers are involved and thus this class of bugs is much less likely to happen.
+
+Another issue with using `math.floor` as a workaround is that code performing a lot of integer calculations is harder to understand, write and maintain.
+
+Especially with applications dealing with pixel graphics, such as 2D games, integer math is so common that `math.floor` could easily become the most commonly used math library function. For these applications, avoiding the calls to `math.floor` is alluring from the performance perspective.
+
+> Non-normative: Here are the top math library functions used by a shipped game that heavily uses Lua: 
+> `floor`: 461 matches, `max`: 224 matches, `sin`: 197 matches, `min`: 195 matches, `clamp`: 171 matches, `cos`: 106 matches, `abs`: 85 matches.
+> The majority of `math.floor` calls disappear from this codebase with the floor division operator.
+
+Finally, Lua has had floor division operator since version 5.3, so its addition to Luau makes it easier to migrate from Lua to Luau and perhaps more importantly use the wide variety of existing Lua libraries in Luau.
+
+## Design
+
+The design mirrors Lua 5.3:
+
+New operators `//` and `//=` will be added to the language. The operator `//` performs division of two operands and rounds the result towards negative infinity. By default, the operator is only defined for numbers. The operator has the same precedence as the normal division operator `/`. `//=` is the compound-assignment operator for floor division, similar to the existing operator `/=`.
+
+A new metamethod `__idiv` will be added. The metamethod is invoked when any operand of floor division is not a number. The metamethod can be used to implement floor division for any user defined data type as well as the built-in vector type.
```

Comment:
Other operators are already built-in for vector type without having to provide a metamethod, why is this operation excluded from the other ones?

Reply:
I guess, similar to how % is not defined and +-*/ are and integer division being a bit strange on vectors, it's ok to skip unless you really want it with a metamethod.

Reply:
Yeah, I also thought floor division on vectors would be quite rare so we could keep the VM minimal by not implementing it in the opcode, at least not in this stage. This can be added easily later if there are significant wins in sight.